File: 3. 无重复字符的最长子串/无重复字符的最长子串.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> charIndex;
        int maxLength = 0;
        int left = 0;

        for (int right = 0; right < s.length(); right++) {
            char currentChar = s[right];

            // 如果字符已存在且在当前窗口内，移动左指针
            if (charIndex.find(currentChar) != charIndex.end() && charIndex[currentChar] >= left) {
                left = charIndex[currentChar] + 1;
            }

            // 更新字符的最新位置
            charIndex[currentChar] = right;

            // 更新最大长度
            maxLength = max(maxLength, right - left + 1);
        }

        return maxLength;
    }
};

int main() {
    Solution solution;

    // 测试用例
    cout << "测试用例 1:" << endl;
    cout << "输入: s = \"abcabcbb\"" << endl;
    cout << "输出: " << solution.lengthOfLongestSubstring("abcabcbb") << endl;
    cout << "预期: 3" << endl << endl;

    cout << "测试用例 2:" << endl;
    cout << "输入: s = \"bbbbb\"" << endl;
    cout << "输出: " << solution.lengthOfLongestSubstring("bbbbb") << endl;
    cout << "预期: 1" << endl << endl;

    cout << "测试用例 3:" << endl;
    cout << "输入: s = \"pwwkew\"" << endl;
    cout << "输出: " << solution.lengthOfLongestSubstring("pwwkew") << endl;
    cout << "预期: 3" << endl << endl;

    cout << "测试用例 4:" << endl;
    cout << "输入: s = \"\"" << endl;
    cout << "输出: " << solution.lengthOfLongestSubstring("") << endl;
    cout << "预期: 0" << endl;

    return 0;
}

File: 3. 无重复字符的最长子串/无重复字符的最长子串.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_index_map={}
        left=0
        max_length=0
        for right in range(len(s)):
            if s[right] in char_index_map and char_index_map[s[right]]>=left:
                left = char_index_map[s[right]]+1
          
            char_index_map[s[right]]=right
            max_length = max(max_length,right-left+1)

        return max_length
    
# 示例用法
if __name__ == "__main__":
    solution = Solution()
    test_string = "abcabcbb"
    result = solution.lengthOfLongestSubstring(test_string)
    print(f"The length of the longest substring without repeating characters in '{test_string}' is: {result}")